Feature/members only

Div685 opened the pull request:
# Members-Only

## About the Project

This is a member's-only project using Ruby on Rails.  This project is a building of an exclusive clubhouse where your members can write anonymous posts. Inside the clubhouse, members can see who the author of a post is but, outside, they can only see the story and wonder who wrote it.

This project uses Devise Gem for User Authentication.

**Devise**
Devise is a flexible authentication solution for Rails based on Warden. It:

- Is Rack based;
- Is a complete MVC solution based on Rails engines;
- Allows you to have multiple models signed in at the same time;
- Is based on a modularity concept: use only what you really need.

Div685 committed:
Add github and rubocop linters

Div685 committed:
Add Application Gem and bootstrap

- Add bootstap and it's setups
- Add Devise Gem and it's setups
- Add simple_form gem
- Add Gravatar Gem

Div685 committed:
Add Post model and controller

- Add index, show, edit, destroy, new method to controller
- Add Post model
- Add db:migrations
- update routes file path to the root
- update index.html.erb to verify bootstrap
- update Applicaiton.css file to .scss
- Add @import 'bootstrap' to application.scss file

Div685 committed:
Add navbar and form

- Add navbar and form parser
- update create method in controller

Div685 committed:
Add user model using Devise

- Add feed, latest-post parser
- update navbar parser
- Add registrations_controller.rb
- Add Association in user and post

Adewunmi97 committed:
add feed and latest post contents

Div685 committed:
Add new.html.erb file

- Add style to sign-up and login page

Adewunmi97 committed:
add readme file

Div685 committed:
Add style to new post

- Add style to feed.html and latest-post
- update navbar

Div685 committed:
Add validation in model

- Add screenshot
- update show.html.erb file
- Add link to show.html.erb file
- add style in posts.scss file

Div685 committed:
update readme.md file

Div685 committed:
update files for linters check

Div685 committed:
update files for linter

Div685 committed:
update posts.scss style for linter

kevcap commented on app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def show; end
+
+  def edit; end
```
Delete those empty methods;

kevcap reviewed (changes requested):
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  your project is looking really good, however, to accomplish as the actual project, you required to do these changes, listed below:

 
## [REQUIRED CHANGES]
- [x] At your PostsController, please note that:
it uses `before_action` to restrict access to the `:new` and `:create` methods (uses ApplicationController helper method) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_action_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))

```ruby
before_action :signed_in_only!, only: [:new, :create]

def signed_in_only!
  # current_user in ApplicationController must be defined as helper method to make it work
  redirect_to  new_user_session_path unless current_user

  ## alternative solution:
  # if signed_in? is defined in ApplicationController (see stretch requirements)
  redirect_to new_user_session_path unless signed_in?
end

```

- [x] Please delete empty methods (show and edit);
- [x] Please remember to only set the routing methods that you'll be using (Create a Post model and a Posts controller and a corresponding resource in your Routes file which allows the [:new, :create, :index] methods.)
## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> <img  alt="gmail"  src="https://img.shields.io/badge/gmail-%20- green?logo=gmail&style=social"></a> [![Twitter Follow](https://img.shields.io/twitter/follow/kevinnnmn?style=social)](https://twitter.com/intent/follow?screen_name=kevinnnmn) [![](https://img.shields.io/badge/Slack-Kevin%20Alves-green?logo=slack&style=social)](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In the real world, you'll have a lot of change requests for your real clients, it's good to get used to it.</sub>

Div685 committed:
update controller and models

- removed empty method
- add routes only to specfied methods
- Add devise strong parameters to application_controller

teekaytech commented on app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def new
+    @post = current_user.posts.build
```
Initializing an empty post with `@post = Post.new` is enough for the `new` action, since `current_user.posts.build` is being used in the `create` action of the post controller.

Div685 replied:
Hello, teekaytech, 
Thanks for the quick review. I can only remove the show from the line of `authenticate_user!` but I can't remove it from routes and `before_action` for a post, It will give me the error because I have used the `show.html.erb` file which uses the link/routes for `post_path(:id)` e.g `<%= link_to post.title, post_path(post) %>`.

teekaytech commented on app/views/posts/_feed.html.erb:
```diff
@@ -0,0 +1,32 @@
+                <div class="content d-flex flex-column">
+                    
+                    <div class="d-flex flex-column m-0 p-0">
+                        <% if user_signed_in? %>
+                            <strong class="text-capitalize "> <%= post.user.name %> </strong>
+                            <small> @<%= post.user.username %> </small>
+                        <% else %>
+                            <strong> Anonymous User </strong>
+                            <small> @Anonymous </small>
+                        <% end %>
```
There should be absolutely NO logic in views. Kindly use the helper file.
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
- [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

teekaytech commented on app/views/posts/_latest-post.html.erb:
```diff
@@ -0,0 +1,31 @@
+            </div>
+            <div class="media-content">
+                <div class="content">
+                <% if user_signed_in? %>
+                        <strong class="text-capitalize "> <%= post.user.name %> </strong><br />
+                        <small> @<%= post.user.username %> </small><br />
+                        <% else %>
+                            <strong> Anonymous User </strong><br />
+                            <small> @Anonymous </small><br />
+                        <% end %>
```
This is another logic. Kindly use the helper for view-related logic.

teekaytech commented on app/views/posts/show.html.erb:
```diff
@@ -0,0 +1,39 @@
+    </div>
+
+    <div class="media-left d-flex">
+                <% if user_signed_in? %>
+                    <figure class="">
+                        <%= gravatar_image_tag(current_user.email,size: 15, alt: current_user.name, class: 'rounded-circle mr-2') %>
+                    </figure>
+                <%else%>
+                    <figure class="">
+                        <%= gravatar_image_tag('link.jpg',size: 15, alt: "current_user", class: 'rounded-circle mr-2') %>
+                    </figure>
+                <% end %>
```
Here's another logic. Kindly move it to the helper file.

teekaytech commented on app/views/posts/index.html.erb:
```diff
@@ -0,0 +1,10 @@
+    <div class="left col-12 p-3">
+        <%= render 'feed' %>    
+    </div>
+    <!-- <div class="right col-3 p-3 card my-3">
+        <%= render 'latest-post' %>
+    </div> -->
```
It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

teekaytech reviewed (changes requested):
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've made laudable progress in this project.
However, some changes are required before it can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] Since you've deleted the show and edit methods, kindly remove the show method from the routes and the `before_action` statement.

- [x] I highlighted in my inline comments below, some instances where you used logic in your view files. Kindly check for all instances where this is happening and encapsulate view-related logic in View helpers (app/helpers/*_helper.rb)



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

Div685 committed:
update post_controller.rb file

Div685 committed:
update index.html.erb file

- remove comment frm file

elmejdki reviewed (changes requested):
# status: require changes :recycle: 

Great job so far guys :clap: :clap: :clap:, but you didn't complete the required changes :hammer: by the previous TSE.

Please see the comment below
:point_down: :point_down: :point_down: :point_down: :point_down:

## Required Changes:

- [x] Please remove all logic from your views they should go inside the view helpers as suggested by the previous TSE.

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat: 
_Feel free to contact me on slack if you have any questions._

Div685 commented:
Hello elmejdki,
I think `user_signed_in?` is already a helper method from Devise and the main purpose of this project is to use Devise and only Authorised user can see the owner of the Post so I'm not so sure how can I make that part.

teekaytech reviewed (changes requested):
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__Everything is looking good. You're almost there!
Let's quickly make this last change. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] You still have some logic in your views. 
- Specifically, view partials (including `_feed.html.erb, _latest.html.erb, _nav.html.erb`) and view files like `show.html.erb` are having logic embedded in them.
- If you're having any difficulty in fixing this, kindly reach out to any of the previous reviewers including me. Moving view-related logic to the helper module is a specific requirement for this project. 



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

Div685 committed:
Add logic to helper class

- update navbar, feed, show, latest-post file
- Add logic to application_helper.rb file

voscarmv reviewed (approved):
# Project status: Approved ✨ 😎👍 💯 ✔️

- Great work applying those last necessary changes.

### You may merge this branch and move forward ⛙🛣

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️
